base/parse.jl

```diff
@@ -38,16 +38,38 @@ julia> parse(Complex{Float64}, "3.2e-1 + 4.5im")
 parse(T::Type, str; base = Int)
 parse(::Type{Union{}}, slurp...; kwargs...) = error("cannot parse a value as Union{}")
 
-function parse(::Type{T}, c::AbstractChar; base::Integer = 10) where T<:Integer
-    a::Int = (base <= 36 ? 10 : 36)
-    2 <= base <= 62 || throw(ArgumentError("invalid base: base must be 2 ≤ base ≤ 62, got $base"))
-    d = '0' <= c <= '9' ? c-'0'    :
-        'A' <= c <= 'Z' ? c-'A'+10 :
-        'a' <= c <= 'z' ? c-'a'+a  : throw(ArgumentError("invalid digit: $(repr(c))"))
-    d < base || throw(ArgumentError("invalid base $base digit $(repr(c))"))
-    convert(T, d)
+@noinline function _invalid_base(base)
+    throw(ArgumentError("invalid base: base must be 2 ≤ base ≤ 62, got $base"))
+end
+
+@noinline _invalid_digit(base, char) = throw(ArgumentError("invalid base $base digit $(repr(char))"))
+
+function parse_char(::Type{T}, c::AbstractChar, base::Integer, throw::Bool) where T
+    a::UInt8 = (base <= 36 ? 10 : 36)
+    (2 <= base <= 62) || _invalid_base(base)
+    base = base % UInt8
+    cp = codepoint(c)
+    cp = cp > 0x7a ? 0xff : cp % UInt8
+    d = UInt8('0') ≤ cp ≤ UInt8('9') ? cp - UInt8('0') :
+        UInt8('A') ≤ cp ≤ UInt8('Z') ? cp - UInt8('A') + UInt8(10) :
+        UInt8('a') ≤ cp ≤ UInt8('z') ? cp - UInt8('a') + a :
+        0xff
+    d < base || (throw ? _invalid_digit(base, c) : return nothing)
+    convert(T, d)::T
+end
+
+function parse(::Type{T}, c::AbstractChar; base::Integer=10) where {T <: Integer}
+    @inline parse_char(T, c, base, true)
```

Comment:
In general, I try to avoid second guessing the compiler on inlining heuristics unless I'm very familiar with the code and all its callsites and/or I've measured performance differences that confirm that I know better than the compiler.

This inlining choice seems accurate to me so I'm willing to proceed with the annotation if you think it helps performance.